### PR TITLE
Implements `Geometry::make_valid`.

### PR DESCRIPTION
**Pull request by metasim**
- [x] I agree to follow the project's [code of conduct](https://github.com/georust/gdal/blob/master/CODE_OF_CONDUCT.md).
- [x] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

Partially addresses #306.

cc: @thomas001

### CODE REVIEW

File: src/metadata.rs [hunk 1]
```diff
@@ -235,7 +235,7 @@ pub trait Metadata: MajorObject {
     }
 }
 
-/// Standalone metadata entry, as returned by iterator from [`Metadata::metadata_iter`].
+/// Standalone metadata entry, as returned by iterator from [`Metadata::metadata`].
```
Comment (metasim) [inline src/metadata.rs:238]: Stray bug.

File: src/vector/geometry.rs [hunk 2]
```diff
@@ -388,21 +389,18 @@ impl Geometry {
...
 
-    /// May or may not contain a reference to a SpatialRef: if not, it returns
-    /// an `Ok(None)`; if it does, it tries to build a SpatialRef. If that
-    /// succeeds, it returns an Ok(Some(SpatialRef)), otherwise, you get the
-    /// Err.
+    /// Get the spatial reference system for this geometry.
     ///
+    /// Returns `Some(SpatialRef)`, or `None` if one isn't defined.
+    ///
+    /// Refer: [OGR_G_GetSpatialReference](https://gdal.org/doxygen/ogr__api_8h.html#abc393e40282eec3801fb4a4abc9e25bf)
```
Comment (metasim) [inline src/vector/geometry.rs:396]: Slipping in documentation synchronization from #355.

File: src/vector/geometry.rs [hunk 3]
```diff
@@ -388,21 +389,18 @@ impl Geometry {
...
         let c_spatial_ref = unsafe { gdal_sys::OGR_G_GetSpatialReference(self.c_geometry()) };
 
         if c_spatial_ref.is_null() {
             None
         } else {
-            match unsafe { SpatialRef::from_c_obj(c_spatial_ref) } {
-                Ok(sr) => Some(sr),
-                Err(_) => None,
-            }
+            unsafe { SpatialRef::from_c_obj(c_spatial_ref) }.ok()
```
Comment (metasim) [inline src/vector/geometry.rs:403]: Consistency.

File: CHANGES.md [hunk 4]
```diff
@@ -22,6 +22,10 @@
 
   - <https://github.com/georust/gdal/pull/355>
 
+- Implemented `Geometry::make_valid` and `Geometry::make_valid_ex`.
```
Comment (metasim) [inline CHANGES.md:25]: When reviewing the following changes, I'm interested in knowing what our policy should be here: 

a. Mirror `foo` & `foo_ex` patterns in C/C++ API
b. Have only `foo`, but always require `FooOpts`, providing an appropriate `FooOpts::default()` to minimize additional ceremony.

Question will make more sense in the context of what is to come.

File: src/metadata.rs (hunk 1 above)
Comment (lnicola) [reply src/metadata.rs:238]: Why can't rustdoc warn about this, I wonder?

File: src/vector/geometry.rs [hunk 5]
```diff
@@ -593,4 +663,38 @@ mod tests {
         // We don't care what it returns when passed an invalid value, just that it doesn't crash.
         geometry_type_to_name(4372521);
     }
+
+    #[test]
+    pub fn test_make_valid() {
```
Comment (metasim) [inline src/vector/geometry.rs:668]: Tests ported from [GDAL](https://github.com/OSGeo/gdal/blob/9f541e8193bd7b1b59ef1aa9ada3bc235fdb56e6/autotest/ogr/ogr_geom.py#L3949).

File: src/vector/geometry.rs [hunk 6]
```diff
@@ -480,9 +514,45 @@ impl Debug for GeometryRef<'_> {
...
+    Linework,
+    /// First makes all rings valid, then merges shells and subtracts holes from shells to generate valid result.
+    ///
+    /// Assumes holes and shells are correctly categorized.
+    Structure {
+        /// If `false`, collapses are converted to empty geometries.
+        /// If `true`, collapses are converted to a valid geometry of lower dimension.
+        keep_collapsed: bool,
+    },
+}
```
Comment (metasim) [inline src/vector/geometry.rs:531]: Purpose here is to try to encode possible options in a way that only expresses valid combination of options. I'd like to think our attempting to provide these ergonomics are in-scope for this library, but I could see arguments against it too.

w.r.t. the opening question: should we just have a single `make_valid` that always requires this (e.g. `geom.make_valid(MakeValidOpts::default())`) or is it better have two separate methods (as this implementation has)?

File: src/vector/geometry.rs [hunk 7]
```diff
@@ -480,9 +514,45 @@ impl Debug for GeometryRef<'_> {
     }
 }
 
+/// Invocation options for [`Geometry::make_valid_ex`].
+#[derive(Debug, Clone, Default)]
+pub enum MakeValidOpts {
```
Comment (lnicola) [inline src/vector/geometry.rs:519]: I feel (not that strongly, though) we should allow the caller to pass in their own options in case a future version of GDAL adds an option we don't know about yet,

File: CHANGES.md (hunk 4 above)
Comment (lnicola) [reply CHANGES.md:25]: I'm inclined to prefer single functions instead of pairs like these. Note that GDAL already has some `Ex` functions, so us making up new ones might end up being quite confusing. But I guess that's not such a great risk, GDAL would have to add a new parameter that can't be expressed as an existing `CSLStringList`.

File: src/vector/geometry.rs (hunk 6 above)
Comment (lnicola) [reply src/vector/geometry.rs:532]: Yeah, but the valid combinations today might be insufficient tomorrow. We could either allow passing a `CSLStringList`, or add a `Custom` variant here -- but then what happens when GDAL adds another unrelated option?

File: src/vector/geometry.rs (hunk 7 above)
Comment (metasim) [reply src/vector/geometry.rs:520]: Thoughts on how that might be reflected in the implementation while not falling back to everything accepting `CslStringList` (which I think misses an opportunity for safer usage).

File: src/metadata.rs (hunk 1 above)
Comment (metasim) [reply src/metadata.rs:238]: It does, if you remember to run it! 😄.

I guess the more specific question is: should we add this as a CI check?

Comment (lnicola) [reply src/metadata.rs:238]: I'd say yes. Are we clean on those warnings now?

File: CHANGES.md (hunk 4 above)
Comment (metasim) [reply CHANGES.md:25]: What if the signature was something like `fn make_valid<O: Into<CslStringList>>(&self, opts: O)`? That way you could use `MakeValidOpts` if you want, or do it all yourself manually? 🤔 

(NB: I think figuring out the right pattern here is important, because looking ahead I see more of this sort of things coming up).

File: src/metadata.rs (hunk 1 above)
Comment (metasim) [reply src/metadata.rs:238]: Yes, should be. Note, I use [this](https://github.com/georust/gdal/blob/master/.cargo/config.toml#L3) (`cargo dto`) when writing docs to help me catch errors in the examples. `rustdoc` isn't enought. Seems we need a combination of that, disallowing `rustdoc` warnings (similar to [this](https://github.com/georust/gdal/blob/master/.cargo/config.toml#L5)).

File: src/vector/geometry.rs (hunk 6 above)
Comment (metasim) [reply src/vector/geometry.rs:532]: How about this approach?

https://github.com/georust/gdal/pull/356#discussion_r1059668374

It's in times like these I wish Rust had default arguments...

File: CHANGES.md (hunk 4 above)
Comment (lnicola) [reply CHANGES.md:25]: Yeah, I thought about that, but it has some disadvantages:

 - it makes the function generic, increasing the compile times and bloating the code if we're not careful to split it in two
 - it's less discoverable, you're ready to use the function, then you realize you have to see what implements `Into<CslStringList>`. Examples can help here and I think rustdoc shows these implementations when you look at the trait, but it's still a bit distracting.

File: src/metadata.rs (hunk 1 above)
Comment (lnicola) [reply src/metadata.rs:238]: This isn't an error in an example, it's a reference to an undefined symbol. But I'll take your word for it :-).

Comment (metasim) [reply src/metadata.rs:238]: Sorry... was conflating documentation error contexts 🤦 . Both are potential issues.

File: CHANGES.md (hunk 4 above)
Comment (metasim) [reply CHANGES.md:25]: @lnicola Would you mind taking a look at this latest version? [Here's how it looks in use](https://github.com/georust/gdal/pull/356/files#diff-4597c3ebc0daa4f8f578c4684bc8c616cb21d6ec4328936ff017c23444d7330cR691-R737).

TBH, because I don't have as much of an aversion to generics/monomorphism (when compared to the benefits), and think documentation can assist with discoverability, I much prefer the `Into<CslStringList>` approach to what is in this version. But I admittedly I have a bias toward ergonomics and have never had to deal with the consequences of code bloat. 

That said, I'll defer to you on what's ultimately best for the project; happy to follow your lead here. 😃 Note: If we *do* go with the generic `CslStringList` parameter version, I'll probably remove `MakeValidOpts` until we come up with a consistent approach to string-list-option construction.

Comment (metasim) [reply CHANGES.md:25]: PS: I'll fix docs and move this out of draft once we settle on a design.

Comment (metasim) [reply CHANGES.md:25]: PS: Just re-read this, and maybe should have drawn attention earlier to the fact that GDAL already has `OGR_G_MakeValid` and `OGR_G_MakeValidEx` variants. So our doubling with `Ex` functions in this case wouldn't technically be making up a new one.

> I'm inclined to prefer single functions instead of pairs like these. Note that GDAL already has some `Ex` functions, so us making up new ones might end up being quite confusing. But I guess that's not such a great risk, GDAL would have to add a new parameter that can't be expressed as an existing `CSLStringList`.

Comment (metasim) [reply CHANGES.md:25]: @lnicola After sleeping on it, I'm now leaning toward this being the API, given the mirroring of `OGR_G_MakeValid`, and `OGR_G_MakeValidEx`.

```
pub fn make_valid(&self) -> Result<Geometry>
```
and
```
pub fn make_valid_ex(&self, opts: &CslStringList) -> Result<Geometry>
```

For now I'll drop `MakeValidOpts`, but possibly consider in the future a separate module of types for idiomatically constructing valid argument combinations and converting them to `CslStringList`, for methods that accept them.

WDYT?

Comment (lnicola) [reply CHANGES.md:25]: Sorry, I'll try to take another look tomorrow.

Comment (metasim) [reply CHANGES.md:25]: > Yeah, but the C++ API only has

huh... you know, until now I just assumed the C API was the core with C++ built on top of that, and not the other way around! 🤦

Comment (lnicola) [reply CHANGES.md:25]: Nope, GDAL is C++, but it exposes a C API. I don't know if the C++ one is actually public.

Comment (metasim) [reply CHANGES.md:25]: @jdroenner @rmanoka @ChristianBeilschmidt or anyone else: do you have opinions about the API approach we should take now an moving forward when optional arguments are specified with `CslStringList`? We're trying to decide a consistent approach as we expose more functions that use a list of `"KEY=VALUE"` strings to specify options.

Comment (jdroenner) [reply CHANGES.md:25]: i think the `O: Into<X> ` pattern is really nice in many cases. However, in this case i think we can go with `Option<X>` or  `CslStringList::default()`. Since it is a list an option is somehow redundant. but maybe more ergonomic.

Comment (lnicola) [reply CHANGES.md:25]: @jdroenner does that trade-off change for you once we bring in `MakeValidOpts`? `make_valid(MakeValidOpts::Structure)` looks a bit nicer than `make_valid(MakeValidOpts::Structure.into())`.

File: src/vector/vector_tests/mod.rs [hunk 8]
```diff
@@ -826,14 +827,17 @@ mod tests {
...
+                let _nolog = SuppressGDALErrorLog::new();
+                // force error
+                assert!(matches!(
+                    layer.set_attribute_filter("foo = bar").unwrap_err(),
+                    GdalError::OgrError {
+                        err: gdal_sys::OGRErr::OGRERR_CORRUPT_DATA,
+                        method_name: "OGR_L_SetAttributeFilter",
+                    }
+                ));
+            }
```
Comment (metasim) [inline src/vector/vector_tests/mod.rs:840]: PR scope creep... I got tired of seeing this error in test output.

File: src/cpl.rs [hunk 9]
```diff
@@ -155,6 +166,47 @@ impl Debug for CslStringList {
...
+///     opts.into().len()
+/// }
+///
+/// assert_eq!(count_opts(&[("One", "1"), ("Two", "2"), ("Three", "3")]), 3);
+/// ```
+impl<const N: usize> From<&[(&str, &str); N]> for CslStringList {
+    fn from(pairs: &[(&str, &str); N]) -> Self {
+        let mut result = Self::default();
+        for (k, v) in pairs {
+            result.set_name_value(k, v).expect("valid key/value pair");
```
Comment (lnicola) [inline src/cpl.rs:204]:
```suggestion
            result.set_name_value(k, v).expect("invalid key/value pair");
```

I think the message would be something like `valid key/value pair: "Invalid characters in name: '#'"`, which doesn't make sense.

And going back a little, I'm not sure what to say about this impl. `from` should be infallible, so `try_from` would be better, which this clearly isn't. On the other hand, it's probably going to be used with compile-constant keys and values. Is that why you've implemented it for fixed-size arrays (as opposed to `&[(&str, &str)]`)?

File: src/vector/geometry.rs [hunk 10]
```diff
@@ -416,6 +454,61 @@ impl Geometry {
...
+    ///
+    /// When GEOS < 3.8, this method will return `Ok(self.clone())` if it is valid, or `Err` if not.
+    ///
+    /// Refer: [OGR_G_MakeValidEx](https://gdal.org/api/vector_c_api.html#_CPPv417OGR_G_MakeValidEx12OGRGeometryH12CSLConstList)
+    ///
+    /// # Example
+    /// ```rust, no_run
+    /// use gdal::vector::Geometry;
+    /// # fn main() -> gdal::errors::Result<()> {
+    /// let src = Geometry::from_wkt("POLYGON ((0 0,10 10,0 10,10 0,0 0))")?;
```
Comment (lnicola) [inline src/vector/geometry.rs:481]: (nit) these are probably nicer to read as below, but you don't need to change them:

```suggestion
    /// let src = Geometry::from_wkt("POLYGON ((0 0, 10 10, 0 10, 10 0, 0 0))")?;
```

**Review by lnicola — CHANGES_REQUESTED** [review]
The error handling changes seem like a really bad idea.

I'm also unsure about the `From` implementation for `CslStringList`, but that's probably less important. From (sic!) [the docs](https://doc.rust-lang.org/std/convert/trait.From.html):

> Note: This trait must not fail. The From trait is intended for perfect conversions. If the conversion can fail or is not perfect, use [TryFrom](https://doc.rust-lang.org/std/convert/trait.TryFrom.html).

Other than that, LGTM.

File: src/cpl.rs (hunk 9 above)
Comment (metasim) [reply src/cpl.rs:204]: > Is that why you've implemented it for fixed-size arrays

No. More of a result of me playing type golf; Without the `const` parameter I get this error. Probably me being dumb.

image


The fact that `set_name_value` is fallible is really annoying, but understandable. Going with `try_from` pretty much eliminates the benefits of having this. So if you're in doubt about it, I think I should just remove it all together.

File: src/vector/layer.rs [hunk 11]
```diff
@@ -251,7 +251,8 @@ pub trait LayerAccess: Sized {
     }
 
     /// Set a feature on this layer layer.
-    /// Refer[SetFeature](https://gdal.org/doxygen/classOGRLayer.html#a681139bfd585b74d7218e51a32144283)
+    ///
+    /// Refer [SetFeature](https://gdal.org/doxygen/classOGRLayer.html#a681139bfd585b74d7218e51a32144283)
```
[the remark this answers is not in the document]
Comment (metasim) [reply src/vector/layer.rs:255]: I don't like `Refer` either... but was trying to be consistent (but forgot the colon here).

**Comment by metasim** [comment]
> The error handling changes seem like a really bad idea.

I detailed my rationale [here](https://github.com/georust/gdal/pull/356#discussion_r1066216433).  Only doing it because a) I believe successful tests should not report error messages, and b) according to the gdal docs, the implementation should be thread-safe.

File: src/vector/geometry.rs [hunk 12]
```diff
@@ -482,13 +575,14 @@ impl Debug for GeometryRef<'_> {
...
+    use super::*;
     use crate::spatial_ref::SpatialRef;
-
-    use super::{geometry_type_to_name, Geometry};
+    use crate::test_utils::SuppressGDALErrorLog;
 
     #[test]
     #[allow(clippy::float_cmp)]
     pub fn test_area() {
+        let _nolog = SuppressGDALErrorLog::new();
```
[the remark this answers is not in the document]
Comment (lnicola) [reply src/vector/geometry.rs:585]: Ah, I missed it being thread-local. Should be fine, then 👍.

**Comment by metasim** [comment]
> I'm also unsure about the `From` implementation for `CslStringList`, but that's probably less important. 

I'm also unsure about this. Maybe I should just drop this `impl` until we figure out a more ergonomic way to construct static `CslStringList`s?

File: src/vector/geometry.rs (hunk 12 above)
Comment (metasim) [inline src/vector/geometry.rs:585]: Without this specific suppression, the log message is

> Warning 1: OGR_G_Area() called against non-surface geometry type.

File: src/cpl.rs (hunk 9 above)
Comment (lnicola) [reply src/cpl.rs:204]: Oof, yes, I think you need to cast it, like in https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0c8e96e6b31a1d7ac28a76cce7abbd1b. Anyway, I suppose it's fine.

**Comment by lnicola** [comment]
We could also implement TryFrom, which seems like a good thing to have anyway.

bors d+

File: src/cpl.rs (hunk 9 above)
Comment (metasim) [reply src/cpl.rs:204]: > I think the message would be something like `valid key/value pair: "Invalid characters in name: '#'"`, which doesn't make sense.


Are you 100% sure? (I'm not) 😄 

Per [the docs](https://doc.rust-lang.org/core/result/enum.Result.html#recommended-message-style)

> We recommend that expect messages are used to describe the reason you expect the Result should be Ok.

**Comment by metasim** [comment]
@lnicola Thanks for slogging through this with me! I'm really happy with the result. Wouldn't be half as good without your feedback! 😃

File: src/cpl.rs (hunk 9 above)
Comment (lnicola) [reply src/cpl.rs:204]: Can you test? 😀 I can't right now, sorry.

My reading of the docs is something like `CslStringList::try_from(&["FOO", "BAR"]).expect("obviously correct")`, that is, user code. It's different in library code which can and will actually panic.

Comment (metasim) [reply src/cpl.rs:204]: I'm not sure I'm following your line of thinking here. Happy to submit a new PR revisiting this.

Comment (lnicola) [reply src/cpl.rs:204]: Indeed, this produces:

```
thread 'cpl::tests::from_impl' panicked at 'valid key/value pair: BadArgument("Invalid characters in name: 'TWO#'")', src/cpl.rs:204:41
```

compared to a random example from `libcore`:

```
self.checked_add(rhs).expect("overflow when adding durations")
```
